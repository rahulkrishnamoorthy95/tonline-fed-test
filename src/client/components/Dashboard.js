import React, { useEffect, useState } from 'react';
import Card from './Card';
import Viewmore from './Viewmore';
import Loading from './Loading';

function Dashboard() {

    const [cardDetails, setDetails] = useState([]);
    const [isLoading, setIsLoading] = useState(true);

    const mapResponseToUI = (response = []) => {
        let result = [];
        const getDate = (dateObj) => {
            let dateString = new Date(dateObj);
            dateString = `${dateString.toLocaleString('default', { month: 'long' })} ${dateString.getDate()}, ${dateString.getFullYear()}`;
            return dateString;
        };
        response.forEach((element, key) => {
            const _UIModel = {};
            const { name, gender, location, email, dob, phone, cell, picture } = element;
            _UIModel.id = key;
            _UIModel.name = name?.first ? `${name.first} ${name.last}` : '';
            _UIModel.gender = gender || '';
            _UIModel.location = location ? `${location.city}, ${location.state}, ${location.country}` : '';
            _UIModel.postcode = location?.postcode;
            _UIModel.email = email || '';
            _UIModel.age = dob?.age || '';
            _UIModel.dob = dob?.date ?  getDate(dob.date): null;
            _UIModel.phone = phone || cell;
            _UIModel.picture = picture?.large || picture?.medium;
            result.push(_UIModel);
        });
        return result;
    }

    const fetchCards = (totalCount) => {
        setIsLoading(true);
        let requestUrl = 'https://randomuser.me/api/';
        if (totalCount) requestUrl = `${requestUrl}?results=${totalCount}`;
        fetch(requestUrl)
        .then(res => res.json())
        .then(response => {
            const result = mapResponseToUI(response.results);
            setDetails([...cardDetails, ...result]);
            setIsLoading(false);
        })
        .catch(error => console.log(error));
    }

    const onViewMoreClick = () => {
        fetchCards(5);
    }

    useEffect(() => {
        fetchCards();
    },[]);


    if (cardDetails.length < 1) {
        return <Loading isLoading={isLoading} />;
    }

    return (
        <section id="content">
            <div className="row">
                {cardDetails.length > 0 && <Card data={cardDetails}/>}
            </div>
            <Viewmore callback={onViewMoreClick} hidden={cardDetails.length < 1}/>
            <div className="total">
                {cardDetails.length > 0 && <span>Total cards: {cardDetails.length}</span>}
            </div>
        </section>
    );

}

export default Dashboard;
