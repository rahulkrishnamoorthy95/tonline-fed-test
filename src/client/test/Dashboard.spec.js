import React from 'react';
import Dashboard from '../components/Dashboard';
import { render, unmountComponentAtNode } from "react-dom";
import { act } from "react-dom/test-utils";

let container = null;
beforeEach(() => {
  container = document.createElement("div");
  document.body.appendChild(container);
});

afterEach(() => {
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe('Loading', () => {
    it("should render initial layout", () => {
        const component = shallow(<Dashboard />);
        expect(component.getElements()).toMatchSnapshot();
    });
  
    it('should initialize cardDetails with []', () => {
        const state = {
            cardDetails: [],
            isLoading: true
        };
        shallow(<Dashboard />);    
        expect(state.cardDetails).toEqual([]);
        expect(state.isLoading).toEqual(true);
    });
    
    it("renders user data", async () => {
        const cards = [{
            name: "Joni Baez",
            age: "32",
            location: "Charming Avenue",
            gender: 'male'
        }];
      
        jest.spyOn(global, "fetch").mockImplementation(() =>
          Promise.resolve({
            json: () => Promise.resolve(cards)
          })
        );
      
        await act(async () => {
          render(<Dashboard  />, container);
        });
      
        expect(container.querySelector("span").textContent).toBe(cards[0].name);
        expect(container.querySelector("cite").textContent).toBe(cards[0].location);
        expect(container.querySelector("div")).toBeInTheDocument();
      
  
        global.fetch.mockRestore();
      });

});