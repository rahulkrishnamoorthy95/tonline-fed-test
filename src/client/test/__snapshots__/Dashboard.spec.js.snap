// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading should render initial layout 1`] = `
Array [
  <Loading
    isLoading={true}
  />,
]
`;
